Couchdb 3326 clustered purge

jiangphcn opened the pull request:
<!-- Thank you for your contribution!

     Please file this form by replacing the Markdown comments
     with your text. If a section needs no action - remove it.

     Also remember, that CouchDB uses the Review-Then-Commit (RTC) model
     of code collaboration. Positive feedback is represented +1 from committers
     and negative is a -1. The -1 also means veto, and needs to be addressed
     to proceed. Once there are no objections, the PR can be merged by a
     CouchDB committer.

     See: http://couchdb.apache.org/bylaws.html#decisions for more info. -->

## Overview

<!-- Please give a short brief for the pull request,
     what problem it solves or how it makes things better. -->

The primary purpose of clustered purge is to clean databases that have multiple deleted tombstones or single documents that contain large numbers of conflicts. But it can also be used to purge any document (deleted or non-deleted) with any number of revisions.

With PR https://github.com/apache/couchdb/pull/1370, the purge API and update on secondary index are implemented. The current PR is in charge of updating geospatial index according to purge tree and record purge history in local purge document to make sure that the index update is not out of date.

## Testing recommendations

<!-- Describe how we can test your changes.
     Does it provides any behaviour that the end users
     could notice? -->
make check skip_deps+=couch_epi apps=hastings

```
======================== EUnit ========================
module 'hastings_bookmark'
module 'hastings_format_legacy_local'
module 'hastings_index_updater'
module 'hastings_purge_docs_test'
  Hastings Purge Tests
Application crypto was left running!
    hastings_purge_docs_test:70: test_purge_single...[0.422 s] ok
    hastings_purge_docs_test:90: test_purge_multiple...[0.257 s] ok
    hastings_purge_docs_test:117: test_purge_multiple2...[0.284 s] ok
    hastings_purge_docs_test:151: test_purge_local_purge_doc...[0.384 s] ok
    hastings_purge_docs_test:224: test_purge_verify_index...[0.215 s] ok
    hastings_purge_docs_test:260: test_purge_remove_local_purge_doc...[0.228 s] ok
    hastings_purge_docs_test:307: test_purge_hook_before_compaction...[0.896 s] ok
[os_mon] cpu supervisor port (cpu_sup): Erlang has closed
    [done in 4.297 s]
  [done in 6.972 s]
module 'hastings_format_view'
module 'hastings_index'
module 'hastings_util'
module 'hastings_fabric'
module 'hastings_httpd_handlers'
module 'hastings_rpc'
hastings_index_test: design_doc_to_index_int_test (module 'hastings_index_test')...[0.072 s] ok
module 'hastings_delete_db_test'
  Hastings Delete Database Tests
    hastings_delete_db_test:65: should_delete_index_after_deleting_database...[0.294 s] ok
    hastings_delete_db_test:104: should_rename_index_after_deleting_database...[0.284 s] ok
erl_child_setup: failed with error 32 on line 240
erl_child_setup: failed with error 32 on line 240
[os_mon] cpu supervisor port (cpu_sup): Erlang has closed
erl_child_setup: failed with error 32 on line 240
erl_child_setup: failed with error 32 on line 240
    [done in 1.247 s]
  [done in 1.381 s]
module 'hastings_plugin_couch_db'
module 'hastings_index_manager'
module 'hastings_fabric_info'
module 'hastings_vacuum'
module 'hastings_format_geojson'
module 'hastings_test_util'
module 'hastings_epi'
module 'hastings_httpd'
module 'hastings_app'
module 'hastings_format'
module 'hastings_httpd_util'
module 'hastings_fabric_search'
module 'hastings_format_legacy'
module 'hastings_sup'
module 'hastings_handle_info_tests'
  hastings handle info test
    hastings_handle_info_tests:52: handle_info_test...[1.137 s] ok
[os_mon] cpu supervisor port (cpu_sup): Erlang has closed
    [done in 1.326 s]
  [done in 1.495 s]
module 'hastings_ken_tests'
  Ken Tests
    hastings_ken_tests:55: ken...[0.130 s] ok
[os_mon] cpu supervisor port (cpu_sup): Erlang has closed
    [done in 0.451 s]
  [done in 0.574 s]
=======================================================
  All 12 tests passed.
==> rel (eunit)
==> db (eunit)
```

## JIRA issue number
https://issues.apache.org/jira/browse/COUCHDB-3326

## Related Issues or Pull Requests

<!-- If your changes affects multiple components in different
     repositories please put links to those issues or pull requests here.  -->

https://github.com/apache/couchdb/pull/1370

## Checklist

- [X] Code is written and works correctly;
- [X] Changes are covered by tests;
- [X] Documentation reflects the changes;

davisp commented on test/hastings_purge_docs_test.erl:
```diff
@@ -0,0 +1,457 @@
+
+setup() ->
+    DbName = ?tempdb(),
+    ok = fabric:create_db(DbName, [?ADMIN_CTX, {q, 1}]),
```
We shouldn't need to go through fabric for these tests.

davisp replied:
We do? Am I forgetting why `hastings_index:search/2` won't work here?

davisp commented on test/hastings_ken_tests.erl:
```diff
@@ -71,6 +71,6 @@ ken(DbName) ->
         ok = meck:reset(hastings_index),
         {ok, _} = fabric:update_doc(DbName, DDoc, [?ADMIN_CTX]),
-        ?assertEqual(ok, meck:wait(8, hastings_index,design_doc_to_indexes, '_', 5000)),
+        ?assertEqual(ok, meck:wait(1, hastings_index,design_doc_to_indexes, '_', 5000)),
```
This change seems odd. Did we change Q somewhere?

davisp commented on src/hastings_vacuum.erl:
```diff
@@ -289,3 +290,27 @@ rename_all_indexes(ShardDbName) ->
+                            ok
+                    end,
+                    couch_db:close(ShardDb)
+                            end, [], LocalShards);
```
Indent here is way off.

davisp commented on src/hastings_util.erl:
```diff
@@ -76,3 +92,116 @@ calculate_delete_directory(DbNameDir) ->
+        lists:map(fun(STIndex) ->
+            maybe_create_local_purge_doc(Db, STIndex)
+        end, STIndexes);
+    true -> ok end.
```
Swap the logic so that we have:

```erlang
if STIndexes == [] -> ok; true ->
    ...
end.
```

davisp commented on src/hastings_index.erl:
```diff
@@ -421,6 +473,40 @@ design_doc_to_index_int(Id, Fields, IndexName) ->
     end.
 
 
+verify_index_exists(Options) ->
```
I'd look more at removing the checks from hastings here. This function returns a true/false so there's not a lot of reason to try and safely handle all of the various errors directly since they'll all lead to the same result for the function.

davisp commented:
I can't figure out how to reply directly to the versioning thing. But we removed the ability to downgrade because its not possible to do that cleanly so we shouldn't need that logic. There's certainly nothing in CouchDB that uses this logic and its been removed from everything in those PRs.

jiangphcn commented:
Hey @davisp I removed versioning in local purge document in commit 58f6aa1 and also added `hastings_plugin_couch_db:get_purge_client_fun/1` for geo index checker for clustered purge. Will work on test.
